ports-of-call/array.hpp

```diff
@@ -0,0 +1,262 @@
+#ifndef _PORTS_OF_CALL_ARRAY_HPP_
+#define _PORTS_OF_CALL_ARRAY_HPP_
+
+#include "portability.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <tuple>
+
+namespace PortsOfCall {
+
+//! The array class is a constexpr (mostly) replacement for std::array
+/*!
+std::array is not fully constexpr until C++20.  Currently we are relying on
+"constexpr"-ness to offload to GPUs.  PortsOfCall::array provides a
+constexpr replacement for std::array with the following notable exceptions:
+    - the reverse iterators are not provided because std::reverse_iterator is
+not constexpr until C++17 (could be provided with modest effort)
+
+    - the `at` function is not provided because it bounds checks and throws an
+exception which we cannot do on a GPU (it could be added and made
+non-constexpr)
```

Comment:
I would suggest providing `at` with `PORTABLE_ALWAYS_THROW_OR_ABORT` which will throw an exception on host, but not device based on the protability strategy. (Requires including `portability.hpp` and `portable_errors.hpp`)